ppy/osu
# Scale tablet output size when UI Scaling mode is "Everything"

Fixes #12098

Required osu-framework changes have been merged in https://github.com/ppy/osu-framework/pull/6460.

## Description

When UI Scaling is set to "Everything", tablet output area will now map to the scaled area instead of the entire display. This resolves some issues discussed in #12098 where full area players who use UI Scaling are unable to use the full tablet area as some of the tablet remained mapped to the unused space outside the UI.

Tablet output will not scale when UI Scaling is set to "Gameplay Only" or "Exclude Overlays" since this results in elements being non-interactable with the tablet, which may or may not be desirable for some users and requires further discussion.

The "Conform to current game aspect ratio" button has also been updated to use the scaled area's aspect ratio. Thanks @timschumi for catching this.

## Demonstration
https://github.com/user-attachments/assets/e7da01bc-bd44-4f8d-bc79-b9ebb59054fa

In the first half of the video I demonstrate how on the current lazer client, when the UI is scaled the tablet area is still mapped to the entire window. This results in a significant portion of the tablet area becoming unusable during gameplay. 

In the second half of the video, I demonstrate how on this branch, when the UI is scaled the tablet area is mapped only to the scaled portion of the game window. This means the full tablet area can be used for gameplay.

## Discussion

When osu! updates to include this change, users who already had UI Scaling set to "Everything" will find that their effective tablet area has increased, since their input area will be the same, while their output area will now map to just the scaled area. This could be a jarring experience for those users.

Is there an existing practice for implementing config migrations? If so, I could write a one-time migration that modifies the user's input area such that their effective area will be unchanged. If not, users will need to manually adjust their area back to their liking. Users who do not use UI Scaling "Everything" will not be impacted.

Alternatively, I could add a checkbox to the settings, so that output area scaling could be toggled. I'd think it should be disabled for existing users, but enabled for new user.

## Manual Testing

**Initial game launch with no configuration (deleted osu-development folder)**
Game initialized with ScalingMode=Off, and tablet set to full area as expected

**Set ScaleMode=Everything**
- Modifed scale
- Modified position
- Resized game window
- Tested in Windowed, Borderless, and Fullscreen

Tablet area appropriately mapped to only the scaled area

**All other ScaleMode values**
Tablet area appropriately mapped to the full screen.

## Automated testing

Testing this might require mocking tablet input in order to ensure the cursor position is correct for the given layout configuration. But I don't see any examples of this in other tests. I'm open to suggestions on how best to introduce automated testing for this feature, as I'm not familiar enough with the existing testing practices.

## Dependencies

Depends on `osu-framework` changes: https://github.com/ppy/osu-framework/pull/6460, which has merged.
